### ISSUE TITLE
DDupl performance windowed games with G-SYNC

### ISSUE THREAD

**Issue by alexdie** [psieg/Lightpack#38]
When using Win8 Desktop Duplication with several games in windowed or borderless windowed mode the games run at 20-25 fps while they are in the foreground. When games are not in focus they runs at normal fps (70-120+) or whatever a background fps limit is set to. 
The issue isn't present when games that support it are in exclusive fullscreen or when using WinAPI as capture source.
Enabling or disabling DX hooking doesn't change the windowed behaviour (and shouldn't?). 

Games I tested so far are Dota 2, Firewatch, World of Warcraft and The Division.

This is on the latest public Windows 10 build with the 64bit version of Lightpack.
Drivers are NVidia 365.19

**Comment by psieg** [comment]
Unfortunately, I can't reproduce your issue. I have ~120 fps in Dota 2 when focused and ~35 when out of focus.
So I'll just throw a bunch of questions at you ;)

* How do you set "a background fps limit"?
* How do you even get Dota to stay at 70 fps when not in focus?
* With foreground, do you mean "has window focus" or "is partially/unobscured visible on screen"?
* Do you run something special or uncommon, like another capturing/streaming software like FRAPS or playclaw, SLI, etc? Some driver power saving features? Anything other than Steam, a browser, maybe skype and a plain Windows?
* Does the framerate go back to normal when you quit Prismatik or only after you restart it?
* I understand the issue does not occur in fullscreen exclusive mode?

**Comment by alexdie** [comment]
I haven't been able to consistently reproduce the issue either. For the past two days everything has been working perfectly, and I haven't been able to figure out what changed. I tried a system restore point back to when the issues were present and the issues didn't reappear. There were several Nvidia driver updates, but I haven't paid close enough attention to be sure one of those fixed it. 

In case it reappears or someone else has a similar issue it I'll provide as much information as I can.

- When talking about background/foreground I am always talking about window focus

- Some games natively support a FPS limit when they are in the background - World of Warcraft for example. In Dota 2 the console command "engine_no_focus_sleep [x]" can be used to limit/unlimit FPS when Dota is not in focus. 

- I have a single GTX 980ti, power saving disabled for VR use. I extensively cleaned out anything that might cause interference, all non-vital background processes and services were closed or stopped and the issue still presisted.

- The framerate went back to normal as soon as I closed Prismatik.

- The issue only happened in windowed (and borderless windowed) modes. Fullscreen exclusive mode was fine for every game that otherwise had issues.


Still, for some reason everything works smoothly now so it's probable that this was an issue on my end.

**Comment by alexdie** [comment]
I'm now able to 100% reproduce the issue. 
It happens when Prismatik is enabled and is using Win8 Desktop Duplication capture, any game is running in windowed mode (or borderless windowed mode) and Nvidia G-SYNC is set to be enabled for windowed games. If all of that is set any game drops to 20-25 FPS. I missed the G-SYNC interaction last time and apparently that setting got changed by a driver update without me noticing.

When I change any of the above the issue disappears.

**Comment by psieg** [comment]
Is there a reason you need windowed mode? If everything works in fullscreen mode, that is definitely the way to go.
Since Prismatik is "just using" the Desktop Duplication API, I'm afraid there isn't much we can do - it's probably a NVIDIA driver / Windows issue...

**Comment by alexdie** [comment]
Borderless windowed mode is a lot more convenient since it doesn't take time to switch between desktop applications and the games. It has become the default in most engines and UWP-based games have moved away from fullscreen entirely. This configuration is a common one for people with a G-SYNC setup. 
That being said, there are multiple workarounds each with at most minor inconvenience.

I tested additional games, and only DirectX games seem to be affected. OpenGL or Vulkan based game engines don't have any issues.

I tried setting the grab interval very low or very high and it had no effect on the refresh rate GSYNC sets. 

I tried to find any additional information on how G-SYNC interacts with this API but there's absolutely nothing. Looking at how the API works there's not much that could be done any differently. There have been similar issues with G-SYNC where it limited game FPS, and those required a driver update to solve.

A driver issue seems likely, I'm just documenting what I found in case someone stumbles upon this.

**Comment by psieg** [comment]
I can reproduce a similar issue without any G-SYNC hardware. Do you have multiple displays and/or do you use the internal GPU of your CPU?

**Comment by alexdie** [comment]
The internal GPU of the CPU is disabled in the UEFI settings. I have tried unplugging my second monitor but the behaviour is the same.

**Comment by hkoivuneva** [comment]
I had the same issue with Firewatch. I fixed it by enabling G-Sync in desktop and fullscreen (nvidia control panel g-sync options), disabling V-Sync in-game AND disabling it through the 3D settings profile for the game. Interesting.

It works great with other games usually, but with some it sometimes goes crazy and the LED fps goes to like 1 frame per 3-5 seconds...How to solve that? I couldn't find anything related to that in Prismatik logs.

**Comment by psieg** [comment]
Same question for you: How many monitors, is the internal GPU active?
If you can, try the sample from [here](https://github.com/psieg/Lightpack/issues/81#issuecomment-253462648) and see if it happens.

My current working theory is it is a problem in the Desktop Duplication API implementation (in Windows or the driver), as prismatik is merely consuming that API...

**Comment by psieg** [comment]
It indeed seems to be a driver / GSync issue.
As far as I understand the driver thinks desktop == lock to 24 FPS for whatever reason.

See https://www.reddit.com/r/Lightpack/comments/5ho2ch/extreme_fps_drop_still_with_unofficial_prismatic/

**Comment by mattkhaw** [comment]
@psieg I know this might be a bit dumb. Is it possible to detect the application whether it is displaying at borderless window / windowed / fullscreen? I noticed that the official Prismatik software is using WinAPI to capture (from my observation) and fullscreen only works with Win8 DDupl. If it is possible to detect, then allow the app to auto switch between it. For now, I'm just switching it manually from the software depending on what application I'm running.

**Comment by psieg** [comment]
What exactly is your problem scenario? AFAIK DDupl ist just fine on Desktop and Fullscreen except you are locked to 24 FPS on Desktop. Is that correct? Does it happen without Prismatik too? How many FPS do you get with WinAPI on Desktop?

**Comment by mattkhaw** [comment]
@psieg Yes, DDupl works fine for both Windowed and FullScreen but the problem is when using Windowed/Borderless, the FPS is locked at 24 FPS while FullScreen works flawlessly. I have to check whether it happens without Prismatik. I didn't test that scenario.

From my observation, Borderless gave me around 60-ish FPS using WinAPI on Resident Evil 7 using 1.2 Resolution Scale. If Resolution Scale is set at 1, then I will get around 100 FPS. I'm on a PG348Q with Gsync enabled for both windowed and fullscreen. For now, I'm toggling between WinAPI and DDupl depending on the mode of the application. It's not a big issue for me. It will be great if there's a way to force the Prismatik app to just autodetect the current active app and toggle between WinAPI and DDupl.

If it can be done, it is just a temporary fix. At least it is something rather than nothing.

**Comment by psieg** [comment]
I see. Is disabling G-SYNC for Desktop use any helpful?

**Comment by mattkhaw** [comment]
Haven't actually tried that yet. Maybe I'll give it a go when I reach home. I'll also try disabling the Prismatik app as well while I'm at it. I'll post my results here later.

**Comment by mattkhaw** [comment]
@psieg I saw the video on the problem. This doesn't happen to me. What should I be looking for when I run the sample app?

**Comment by psieg** [comment]
It should cause the same problems as Prismatik, aka if you run the game in windowed mode it should lock to 24FPS. If it does not, Prismatik does something different that causes the problem. If it does, it's a driver issue and there is nothing we can do...

**Comment by mattkhaw** [comment]
The reason why I suggested this is because of the behavior on the official Prismatik software. I guess if there's nothing we can do about it, then I'll just toggle the modes manually depending on what I'm running. I'll check it out and post my results here when I get home.

**Comment by mattkhaw** [comment]
@psieg I've tested according to what you've stated. Here's a rundown.

- G-SYNC disabled with Prismatik running. No issues here.
- G-SYNC enabled with Prismatik disabled. No issues here either.
- G-SYNC enabled with Prismatik running. 24 FPS lock on borderless window/windowed.

I've tried your desktop duplication example. Seems to be working fine with G-SYNC. When the app is maximized, the FPS dipped to around 40 FPS (not locked) but I think that is expected because the program duplicates the screen to no end. I guess there's some sort of performance hit. Once the window is restored to normal size, the FPS goes back up.

I think there's something wrong in Prismatik which is causing the issue, which is why I suggested if possible to detect the window state of the app which is running, then set the mode accordingly. From what I can see, WinAPI works flawlessly for borderless / windowed while DDupl works flawlessly for fullscreen. I've been toggling the official and unofficial software. I think the reason why the official software doesn't work on fullscreen is because it is using WinAPI. I could be wrong.

I do hope I can be of some use but sadly I'm not very well versed with Qt. If I have some time, I might pick up on it and see what I can do to help trace the problem.

**Comment by psieg** [comment]
**Okay, please confirm: If you run your game while the sample application runs, the game has >24 FPS?**

That would mean Prismatik handles DDupl different from the sample, so it would be a bug I can hopefully fix.
WinAPI usually has a high performance impact, so what you suggest would be a workaround - I'd rather fix the problem...

Note: You can also toggle the grab method in the unofficial one instead of switching the entire program

**Comment by mattkhaw** [comment]
Yes. I get more than 24 FPS in game when running the sample application in the background. I've closed Prismatik before I opened the sample application.

Yeah, Prismatik is doing something weird in DDupl. Maybe doing something extra.

Well, that's why I suggest a workaround for the time being until the issue is resolved but this workaround can be applied manually anyways. Just have to remember to select the correct grab mode before running any application. Not an issue for me.

**Comment by hkoivuneva** [comment]
Yeah, with G-Sync ON for Desktop & full screen, FPS drops to 24 with this Prismatik running. 
If I change G-Sync to full screen only, no FPS drop with this Prismatik running.
If I have desktop & full screen G-Sync on and this unofficial Prismatik closed, no FPS drop. 

It happens with both WinApi and DDupl in windowed & borderless windowed mode. Also one weird thing... When the game (WoW 4.3.4) is focused in windowed mode (when I'm interacting with it) FPS drops to 24. When it's out of focus, like now typing here, FPS is very high.  

FPS drops in-game with and without the sample app running when Prismatik is on. When I turn Prismatik off, FPS returns to normal.

Also (game in windowed mode):
1. I turned the game on with G-Sync on in desktop & full screen.
2. When game is clicked, FPS locks to 60 (max refresh rate).
3. When game is out of focus, FPS races high up (game has v-sync disabled).
4. When I start the DDupl sample app while game is running, then click the game, FPS drops to 24. Out of focus FPS is similar to when the app is not running.
5. Then I closed the DDupl app, started Prismatik with WinApi, and FPS was locked to 60 when in focus and went up when not in focus.
6. Started DDupl app: FPS to 24, out of focus still high.
7. Closed DDupl app, FPS goes back to 60 (max refresh rate).
8. Switched Prismatik to DDupl, FPS goes down to 24.
9. Switched back to WinApi, FPS stays at 24. 
10. Closed Prismatik, started game, then started Prismatik again with WinApi on from previous session. FPS locked to 60 while in focus, unlocked while out of focus.
11. Closed game, switched G-Sync to full screen only, started game with Prismatik in DDupl and FPS is unlocked, same with WinApi. 
12. With G-Sync in full screen only, I closed Prismatik, started the sample app, started the game, and FPS is high. Then I closed the game, switched G-Sync to full screen & desktop, started the game with sample app running in the background, and now FPS is again locked to 24.

So could it be the way desktop duplication & G-Sync together are handled in Windows with NVIDIA drivers? If so, we might be out of luck for G-Sync in desktop & full screen mode...

But it works great in full screen games and unless FPS goes to around 40 or lower in borderless/windowed mode and starts to stutter badly, V-Sync isn't too bad to watch! Luckily none of the games I got that can't run in real full screen don't struggle with FPS so it's not a problem.

THANK YOU @psieg for working on this great app and helping us enjoy breathtaking gaming experiences.

Note: Running with only one monitor
![img_20170416_223048](https://cloud.githubusercontent.com/assets/554408/25617452/0f233d14-2f4b-11e7-972a-39cd8eb6f97a.jpg)

**Comment by psieg** [comment]
So @mattkhaw can get more than 24FPS with the sample and @hkoivuneva can not?
@mattkhaw can you please repeat the steps described by @hkoivuneva  and verrify?

**Comment by mattkhaw** [comment]
@psieg I'll try his steps when I get home. I haven't tested in detail like this. From what I experienced, changing from DDupl to WinAPI when game is running won't restore FPS. It will get stuck at 24 no matter what. I don't know if this is related or not, I'm running the sample DDupl app only, not with Prismatik running simultaneously then I have a variable framerate and yes, it is over 24 FPS.

I always start the game fresh and not toggling between sessions like this because from a developer's POV, some APIs which are usually being used will not release the instance as long as the app is running, which is why I always restart the game before changing any specific configurations or running any background apps.

I'm also running the latest drivers from NVIDIA for my PC right now. Windows 10 is the latest version but before the Windows Creator update. Just FYI.

**Comment by hkoivuneva** [comment]
It'll be interesting to hear how it works for you that way! Each machine is different... I'll update to the latest drivers this week/weekend and try again. I haven't yet received Creator's update from Microsoft either.To clarify, I tested it the same way, restarted games in between, because I also noticed that was needed.8.5.2017 4.06 mattkhaw <notifications@github.com> kirjoitti:@psieg I'll try his steps when I get home. I haven't tested in detail like this. From what I experienced, changing from DDupl to WinAPI when game is running won't restore FPS. It will get stuck at 24 no matter what. I don't know if this is related or not, I'm running the sample DDupl app only, not with Prismatik running simultaneously then I have a variable framerate and yes, it is over 24 FPS.
I always start the game fresh and not toggling between sessions like this because from a developer's POV, some APIs which are usually being used will not release the instance as long as the app is running, which is why I always restart the game before changing any specific configurations or running any background apps.
I'm also running the latest drivers from NVIDIA for my PC right now. Windows 10 is the latest version but before the Windows Creator update. Just FYI.

—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or mute the thread.


  
  




8.5.2017 4.06 mattkhaw <notifications@github.com> kirjoitti:@psieg I'll try his steps when I get home. I haven't tested in detail like this. From what I experienced, changing from DDupl to WinAPI when game is running won't restore FPS. It will get stuck at 24 no matter what. I don't know if this is related or not, I'm running the sample DDupl app only, not with Prismatik running simultaneously then I have a variable framerate and yes, it is over 24 FPS.
I always start the game fresh and not toggling between sessions like this because from a developer's POV, some APIs which are usually being used will not release the instance as long as the app is running, which is why I always restart the game before changing any specific configurations or running any background apps.
I'm also running the latest drivers from NVIDIA for my PC right now. Windows 10 is the latest version but before the Windows Creator update. Just FYI.

**Comment by mattkhaw** [comment]
@hkoivuneva I see. Well, I'll try with your testing method and will report the results later. I also haven't received the Creator's Update which is why I mentioned that my copy of Win 10 is updated to the latest version without it just to rule out the possibility of that being a factor. Just hope that this issue doesn't differ from each machine otherwise, there's no way to find the root of the problem.

**Comment by mattkhaw** [comment]
@psieg Okay, I've tested the scenarios which @hkoivuneva stated. It turns out that the internal FPS counter is locked to 24 no matter what. The > 24 FPS thing is actually my display's FPS counter. I think it is trying to match the current framerate, that's why it is constantly running up and down. As for internal FPS counter, I used NVIDIA's one. I didn't bother to install others. So, as for the testing, I got similar results as @hkoivuneva.

From what I see, if you disable G-Sync for windowed mode, it's not going to be a problem. I think most likely G-Sync is an issue for windowed since full screen mode doesn't have this problem. I've tried the sample DDupl app. It has the same results as the Prismatik one. I've also tried activating both for the fun of it to see whether the FPS will tank or not. Obviously, it is locked to 24 FPS even running both.

FYI, I'm using the latest drivers from NVIDIA (382.05). I guess from these testings, we can conclude that there's nothing to do with Prismatik afterall. Just a conflict between G-Sync and DDupl unless you guys are able to find faults in the source code. The only way to solve this is to follow my suggestion. If want to play in windowed / borderless, just use WinAPI, otherwise use DDupl.
